### ISSUE TITLE
[CLI] Extend training support to all trainers

### ISSUE THREAD

**Issue by lewtun** [huggingface/trl#2101]
### Feature request

The CLI currently supports training models with SFT/DPO/KTO: https://github.com/huggingface/trl/blob/6859e048da601fec181997a324e7b351fc997a33/trl/commands/cli.py#L24

It would be good to extend this support to all trainers so that we have a consistent API and also learn which parts of our scripts need refactoring to support this usage.

This could be tackled in separate PRs to keep things lightweight, and I'll track here the trainers in terms of priority to add (based on Hub usage):

* [orpo](https://huggingface.co/docs/trl/orpo_trainer)
* [reward modelling](https://huggingface.co/docs/trl/reward_trainer)
* [ppo](https://huggingface.co/docs/trl/ppov2_trainer)
* [rloo](https://huggingface.co/docs/trl/rloo_trainer)
* [online dpo](https://huggingface.co/docs/trl/online_dpo_trainer)
* [gkd](https://huggingface.co/docs/trl/gkd_trainer)
* [cpo](https://huggingface.co/docs/trl/cpo_trainer)
* [bco](https://huggingface.co/docs/trl/bco_trainer)
* [xpo](https://huggingface.co/docs/trl/xpo_trainer)
* [nash-md](https://huggingface.co/docs/trl/nash_md_trainer)

### Motivation

It is somewhat annoying that one cannot train a model through the CLI as this is helpful for fast debugging / iterations.

### Your contribution

Happy to open PRs, but this could be a good first issue for new contributors!

**Comment by qgallouedec** [comment]
Duplicate #1811, closing it in favour of this one

**Comment by grumpyp** [comment]
Hi, it looks like it'd be just the extension of the `SUPPORTED_COMMANDS` constant?

as for instance `orpo` would already be there https://github.com/huggingface/trl/blob/main/examples/scripts/orpo.py

I didn't check for the others but compared the scripts of `kto` with `orpo` for instance.

**Comment by lewtun** [comment]
@grumpyp yes I think this is all one needs - a better solution would be to have an automated way to populate this constant by globbing all the `_trainer.py` files (e.g. `make cli_commands`). This way, anytime someone adds a new trainer, we automatically get support for it in the CLI :)

**Comment by grumpyp** [comment]
> @grumpyp yes I think this is all one needs - a better solution would be to have an automated way to populate this constant by globbing all the `_trainer.py` files (e.g. `make cli_commands`). This way, anytime someone adds a new trainer, we automatically get support for it in the CLI :)

yes definitely! If you want, assign the issue to me please. I'll try to get a PR out today.

**Comment by lewtun** [comment]
assigned! thanks for the offer to help 🤗

**Comment by grumpyp** [comment]
hi @lewtun

I didn't want to manipulate `py`-files via the Makefile so I went a slighly different approach.

It now creates the commands dynamically using a utility function which is cached. 

Let me know if that works for you or if it needs changes. Thanks for the opportunity to contribute.

**Comment by rp440** [comment]
Looks like this issue should be marked as closed.

**Comment by sezan92** [comment]
is this issue solved, if not i want to start contributing to trl by working on it

**Comment by qgallouedec** [comment]
Thanks! Your contribution is welcome!

**Comment by murilo-cunha** [comment]
Hi @qgallouedec - is this issue still relevant?
Would be happy to contribute! I also have PR #4757 open. I'd appreciate it if someone could take a look when they have a moment.

**Comment by albertvillanova** [comment]
I think we should implement only non-experimental trainers. What do you think?

**Comment by murilo-cunha** [comment]
@albertvillanova not sure the question is directed at me 😅 - but you mean so that we only expose "stable" trainers?

Could make sense, though currently KTO trainer is experimental, and if we only include non-experimental trainers, I don't think this issue is still relevant? All the [not yet implemented trainers are experimental](https://github.com/huggingface/trl/issues/2101#issue-2542170367) if I'm not mistaken?

Also tagging @qgallouedec and @lewtun since they created/edited the issue 😇 

Thanks for the response!

**Comment by qgallouedec** [comment]
Agree with @albertvillanova, 

> but you mean so that we only expose "stable" trainers?

those that are not in trl.experiemental, ie SFT, DPO, Reward, KTO and RLOO and GRPO.

consequently this issue can be closed
